File: src/utils/Logger.ts
import * as core from '@actions/core';

/**
 * Logger utility for consistent logging throughout the application
 */
export class Logger {
  private static instance: Logger;

  constructor() {
    if (Logger.instance) {
      return Logger.instance;
    }
    Logger.instance = this;
  }

  /**
   * Log an info message
   */
  info(message: string): void {
    core.info(message);
    console.log(`[INFO] ${new Date().toISOString()}: ${message}`);
  }

  /**
   * Log a warning message
   */
  warn(message: string): void {
    core.warning(message);
    console.warn(`[WARN] ${new Date().toISOString()}: ${message}`);
  }

  /**
   * Log an error message
   */
  error(message: string): void {
    core.error(message);
    console.error(`[ERROR] ${new Date().toISOString()}: ${message}`);
  }

  /**
   * Log a debug message
   */
  debug(message: string): void {
    core.debug(message);
    console.debug(`[DEBUG] ${new Date().toISOString()}: ${message}`);
  }

  /**
   * Log a notice message
   */
  notice(message: string): void {
    core.notice(message);
    console.log(`[NOTICE] ${new Date().toISOString()}: ${message}`);
  }
}

File: src/ghas_reporting_engine/utils/logger.py
"""
Logging configuration for GHAS Reporting Engine.

This module provides centralized logging configuration for the application.
"""

import logging
import sys
from typing import Optional


def setup_logging(level: str = "INFO", format_string: Optional[str] = None) -> None:
    """Set up logging configuration."""

    # Convert string level to logging constant
    numeric_level = getattr(logging, level.upper(), logging.INFO)

    # Default format
    if format_string is None:
        format_string = "%(asctime)s - %(name)s - %(levelname)s - %(message)s"

    # Configure root logger
    logging.basicConfig(
        level=numeric_level,
        format=format_string,
        handlers=[logging.StreamHandler(sys.stderr)],
    )

    # Set specific loggers to appropriate levels
    logging.getLogger("urllib3").setLevel(logging.WARNING)
    logging.getLogger("requests").setLevel(logging.WARNING)

    # Create main logger
    logger = logging.getLogger("ghas_reporting_engine")
    logger.setLevel(numeric_level)

    logger.debug(f"Logging configured at {level} level")


def get_logger(name: str) -> logging.Logger:
    """Get a logger instance."""
    return logging.getLogger(f"ghas_reporting_engine.{name}")
